### DISCUSSION TITLE
HttpClient error

### DISCUSSION THREAD

**Discussion by callumeveratt** [nanoframework/Home#1365] [Q&A]
### Library/API/IoT binding

nanoframework.System.Net.Http

### Visual Studio version

N/A

### .NET nanoFramework extension version

N/A

### Target name(s)

ESP32 (ESP32-D0WDQ6 (revision v1.0))

### Firmware version

1.8.1.548

### Device capabilities

Connected to: 
ESP32 (ESP32-D0WDQ6 (revision v1.0))
Features WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Flash size 4MB unknown from  (manufacturer 0x104 device 0x16406)
PSRAM: not available
Crystal 40MHz

### Description

We have a while loop which should run a get request using a HttpClient every 15 minutes. Each time the first request works, then each subsequent request will fail with the below error. I thought this may be related to #1339 but I am running the latest version of the nuget package.

Booting up 
API Responded: true
-- NEW REQUEST --
    ++++ Exception System.Net.Sockets.SocketException - CLR_E_FAIL (1) ++++
    ++++ Message: 
    ++++ System.Net.Security.SslNative::SecureClientInit [IP: 0000] ++++
    ++++ System.Net.Security.SslStream::Authenticate [IP: 0045] ++++
    ++++ System.Net.Security.SslStream::AuthenticateAsClient [IP: 000a] ++++
    ++++ System.Net.HttpWebRequest::EstablishConnection [IP: 027c] ++++
    ++++ System.Net.HttpWebRequest::SubmitRequest [IP: 001a] ++++
    ++++ System.Net.HttpWebRequest::GetResponse [IP: 000c] ++++
    ++++ System.Net.Http.HttpClientHandler::Send [IP: 00a7] ++++
    ++++ System.Net.Http.HttpMessageInvoker::Send [IP: 0013] ++++
    ++++ System.Net.Http.HttpClient::Send [IP: 006f] ++++
    ++++ System.Net.Http.HttpClient::Get [IP: 000c] ++++
    ++++ WheelieWatch.Device.Services.HttpRequest::CheckRequest [IP: 002c] ++++
    ++++ WheelieWatch.Device.Program::Main [IP: 00ba] ++++
    ++++ Exception System.Net.WebException - 0x00000000 (1) ++++
    ++++ Message: GetResponse() failed
    ++++ System.Net.HttpWebRequest::GetResponse [IP: 00da] ++++
    ++++ System.Net.Http.HttpClientHandler::Send [IP: 00a7] ++++
    ++++ System.Net.Http.HttpMessageInvoker::Send [IP: 0013] ++++
    ++++ System.Net.Http.HttpClient::Send [IP: 006f] ++++
    ++++ System.Net.Http.HttpClient::Get [IP: 000c] ++++
    ++++ WheelieWatch.Device.Services.HttpRequest::CheckRequest [IP: 002c] ++++
    ++++ WheelieWatch.Device.Program::Main [IP: 00ba] ++++
    ++++ Exception System.Net.Http.HttpRequestException - 0x00000000 (1) ++++
    ++++ Message: An error occurred while sending the request
    ++++ System.Net.Http.HttpClientHandler::Send [IP: 00bf] ++++
    ++++ System.Net.Http.HttpMessageInvoker::Send [IP: 0013] ++++
    ++++ System.Net.Http.HttpClient::Send [IP: 006f] ++++
    ++++ System.Net.Http.HttpClient::Get [IP: 000c] ++++
    ++++ WheelieWatch.Device.Services.HttpRequest::CheckRequest [IP: 002c] ++++
    ++++ WheelieWatch.Device.Program::Main [IP: 00ba] ++++
Exception thrown: 'System.Net.Http.HttpRequestException' in System.Net.Http.dll
An unhandled exception of type 'System.Net.Http.HttpRequestException' occurred in System.Net.Http.dll
Additional information: An error occurred while sending the request

### Sample project or code

// Program.cs
<code>while (true) 
					{
						HttpRequest.CheckRequest();
						Thread.Sleep(900000);
					}</code>

// HttpRequest.CheckRequest
<code>HttpClient HttpClient = new HttpClient();
			HttpClient.HttpsAuthentCert = new X509Certificate(@"-----CERTIFICATE REMOVED FOR BREVITY-----");
		var url = "https://localhost:8080/CheckDay";
			var response = HttpClient.Get(url);
			var responseBody = response.Content.ReadAsString();
			Debug.WriteLine("API Responded: " + responseBody);</code>

**Comment by josesimoes** [comment]
What device are you connecting to?

**Comment by josesimoes** [comment]
Also, have you tried to add the option to close the socket after the request?

**Comment by torbacz** [comment]
Also, have you tried to keep only one HttpClient instead of creating it each time?

**Comment by josesimoes** [comment]
Nice catch. I overlooked that when looking at the code. In fact, that is the intended usage pattern!

**Comment by callumeveratt** [comment]
Thanks both - tried the HttpClient.Dispose() on the original code - same issue still.

I also started with a single usage httpClient but I hit a different error with that - I may be missing something though:
<code>
		static HttpClient _httpClient = new HttpClient()
		{
			HttpsAuthentCert = new X509Certificate(@"CERT FOR BREVITY")
	};
</code>

    ++++ Exception System.NullReferenceException - CLR_E_NULL_REFERENCE (2) ++++
    ++++ Message: 
    ++++ System.Net.WebHeaderCollection::GetValues [IP: 0007] ++++
    ++++ System.Net.Http.Headers.HttpRequestHeaders::set_ConnectionClose [IP: 000a] ++++
    ++++ WheelieWatch.Device.Services.HttpRequest::.cctor [IP: 0003] ++++
Exception thrown: 'System.NullReferenceException' in System.Net.Http.dll
An unhandled exception of type 'System.NullReferenceException' occurred in System.Net.Http.dll

**Comment by torbacz** [comment]
Just create it before while loop and pass to CheckRequest method.

**Comment by callumeveratt** [comment]
I am creating it outside the while loop - it's created within the HttpRequest class at the moment. I have tried it within the Program class too - and same error.

**Comment by torbacz** [comment]
According to your code snipper, you are creating it each time method is executed.
![image](https://github.com/nanoframework/Home/assets/19893965/b4d1dc93-7588-4776-9cb1-c3244e9cceee)

**Comment by callumeveratt** [comment]
So I had that in originally yes, as creating a single instance actually brings back a different error. 



> Thanks both - tried the HttpClient.Dispose() on the original code - same issue still.
> 
> I also started with a single usage httpClient but I hit a different error with that - I may be missing something though: ` static HttpClient _httpClient = new HttpClient() { HttpsAuthentCert = new X509Certificate(@"CERT FOR BREVITY") }; `
> 
> ```
> ++++ Exception System.NullReferenceException - CLR_E_NULL_REFERENCE (2) ++++
> ++++ Message: 
> ++++ System.Net.WebHeaderCollection::GetValues [IP: 0007] ++++
> ++++ System.Net.Http.Headers.HttpRequestHeaders::set_ConnectionClose [IP: 000a] ++++
> ++++ WheelieWatch.Device.Services.HttpRequest::.cctor [IP: 0003] ++++
> ```
> 
> Exception thrown: 'System.NullReferenceException' in System.Net.Http.dll An unhandled exception of type 'System.NullReferenceException' occurred in System.Net.Http.dll

**Comment by torbacz** [comment]
Can you post current code?

**Comment by callumeveratt** [comment]
Okay - so I've changed it to a single httpClient, as suggested. This is my HttpRequestService.cs file.

```
namespace CE.Device.Services
{
	public static class HttpRequest
	{
		static HttpClient _httpClient = new HttpClient()
		{
			HttpsAuthentCert = new X509Certificate(@"--------")
		};

		public static void CheckRequest()
		{
			var url = "https://localhost/CheckDate";
			var response = _httpClient.Get(url);
			response.EnsureSuccessStatusCode();
			var responseBody = response.Content.ReadAsString();
			Debug.WriteLine("API Responded: " + responseBody);
		}
	}
}
```

Then called via Program.cs, unchanged:
```
while (true)
{
	HttpRequest.CheckRequest();
	Thread.Sleep(60000);
}
```

The error now:
++++ Exception System.NullReferenceException - CLR_E_NULL_REFERENCE (2) ++++
++++ Message: 
++++ System.Net.WebHeaderCollection::GetValues [IP: 0007] ++++
++++ System.Net.Http.Headers.HttpRequestHeaders::set_ConnectionClose [IP: 000a] ++++
++++ WheelieWatch.Device.Services.HttpRequest::.cctor [IP: 0003] ++++

**Reply by torbacz** [reply]
I would:
- create HttpClient outside of while loop (it's not the case now)
- then change CheckRequest method to CheckRequest(HttpClient client)
- get rid of static _httpClient

Also I don't see how HttpRequest is created and from exception it seems your header collection may be empty.

**Reply by callumeveratt** [reply]
Okay so I've swapped to this - figured it may be the best way to handle it for now but still getting the same error (I think) the debugger isn't a massive fan of when it blows out with this error - and it BSOD's my laptop.

**Program.cs**
```
namespace CE.Device
{
	public class Program
	{
		public static void Main()
		{
			Debug.WriteLine("Booting up");
			WifiNetworkHelper.Reconnect(requiresDateTime: true);
			HttpClient client = new HttpClient();
			client.HttpsAuthentCert = new X509Certificate(@"--- REMOVED FOR BREVITY ---");

			while (true)
			{
				HttpRequest.CheckRequest(client);
				Thread.Sleep(60000);
			}
		}
	}
}
```

**HttpRequestService.cs**
```
namespace CE.Device.Services
{
	public static class HttpRequest
	{
		public static void CheckRequest(HttpClient client)
		{
			var url = "https://api.com/CheckDate";
			var response = client.Get(url);
			response.EnsureSuccessStatusCode();
			var responseBody = response.Content.ReadAsString();
			Debug.WriteLine("API Responded: " + responseBody);
		}
	}
}
```

Per before - the first request responds (true) as expected. The next request then gives the error. About to try get the message and make sure it's still the same error.

**Reply by callumeveratt** [reply]
Yeah - I managed to see the error. Same error as before. First request good, subsequent requests bad

**Reply by callumeveratt** [reply]
Resumed looking at this last night and I'm now getting this error:
    ++++ Exception System.Net.Sockets.SocketException - CLR_E_FAIL (1) ++++
    ++++ Message: 
    ++++ System.Net.Security.SslNative::SecureClientInit [IP: 0000] ++++
    ++++ System.Net.Security.SslStream::Authenticate [IP: 0045] ++++
    ++++ System.Net.Security.SslStream::AuthenticateAsClient [IP: 000a] ++++
    ++++ System.Net.HttpWebRequest::EstablishConnection [IP: 027c] ++++
    ++++ System.Net.HttpWebRequest::SubmitRequest [IP: 001a] ++++
    ++++ System.Net.HttpWebRequest::GetResponse [IP: 000c] ++++
    ++++ System.Net.Http.HttpClientHandler::Send [IP: 00a7] ++++
    ++++ System.Net.Http.HttpMessageInvoker::Send [IP: 0013] ++++
    ++++ System.Net.Http.HttpClient::Send [IP: 006f] ++++
    ++++ System.Net.Http.HttpClient::Get [IP: 000c] ++++
    ++++ WheelieWatch.Device.Services.HttpRequest::CheckRequest [IP: 001d] ++++
    ++++ WheelieWatch.Device.Program::Main [IP: 00d6] ++++

**Comment by josesimoes** [comment]
I'm moving this to discussions, as this doesn't look like a bug. 
If it is, then it can be moved back to issues.

**Comment by josesimoes** [comment]
@callumeveratt you're calling "localhost", I presume that's the same device that you're making the request from.
Is this correct?
You still haven't reply to my question about the platform that you're making the request to.

**Reply by callumeveratt** [reply]
Sorry - I should have probably just put a better placeholder in there. It's actually an API that I own on the WWW. That all responds as it should on first request - just this second request issue.

**Reply by alberk8** [reply]
Try add the reponse.Dispose();

```
public static void CheckRequest(HttpClient client)
{
	var url = "https://api.com/CheckDate";
	var response = client.Get(url);
	response.EnsureSuccessStatusCode();
	var responseBody = response.Content.ReadAsString();
	Debug.WriteLine("API Responded: " + responseBody);
         response.Dispose();
}
```

**Reply by callumeveratt** [reply]
Tried that originally - same error still from the second request. 

Object definitely gets disposed properly too - as if I try call it again after then I get an ObjectDisposedException

**Reply by josesimoes** [reply]
Can you try debugging the HTTP library?
You have to remove NuGet reference, clone HTTP repo, add the project to your solution and set breakpoints as usual.

**Reply by alberk8** [reply]
Can you try this [sample](https://github.com/alberk8/nanoFramework-HttpClient)

**Reply by callumeveratt** [reply]
@josesimoes @alberk8 Thanks both, will try both things on Sunday when I'm back. Apologies as been away.

**Reply by callumeveratt** [reply]
@alberk8 Think I've got my code working by amending to work in a similar way to yours. Thank you :)
